crates/driver/src/logic/eth/mod.rs

```diff
@@ -54,8 +58,82 @@ impl From<H160> for Address {
     }
 }
 
+impl From<Address> for H160 {
+    fn from(address: Address) -> Self {
+        address.0
+    }
+}
+
+/// A contract address.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Contract(pub H160);
+
+impl From<H160> for Contract {
+    fn from(inner: H160) -> Self {
+        Self(inner.into())
+    }
+}
+
+impl From<Contract> for ethereum_types::H160 {
+    fn from(contract: Contract) -> Self {
+        contract.0 .0.into()
+    }
+}
+
+/// The contract is an address on the blockchain.
+impl From<Contract> for Address {
+    fn from(contract: Contract) -> Self {
+        contract.0.into()
+    }
+}
+
 /// An ERC20 token.
 ///
 /// https://eips.ethereum.org/EIPS/eip-20
 #[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
 pub struct Token(pub H160);
+
+impl From<Token> for H160 {
+    fn from(token: Token) -> Self {
+        token.0
+    }
+}
+
+/// An asset on the Ethereum blockchain. Represents a particular amount of a
+/// particular token.
+#[derive(Debug, Clone, Copy)]
+pub struct Asset {
+    pub amount: U256,
+    pub token: Token,
+}
+
+// TODO Do some web browsing to improve this doc comment.
+/// An amount of native Ether tokens.
+#[derive(Debug, Clone, Copy)]
+pub struct Ether(pub U256);
+
+impl From<Ether> for num::BigInt {
+    fn from(ether: Ether) -> Self {
+        let mut bytes = [0; 32];
+        ether.0.to_big_endian(&mut bytes);
+        num::BigUint::from_bytes_be(&bytes).into()
+    }
+}
+
+impl From<Ether> for U256 {
+    fn from(ether: Ether) -> Self {
+        ether.0
+    }
+}
+
+/// Block number.
+#[derive(Debug, Clone, Copy)]
+pub struct BlockNo(pub u64);
+
+/// An onchain transaction which interacts with a smart contract.
+#[derive(Debug)]
+pub struct Interaction {
+    pub target: Address,
+    pub value: U256,
```

Comment:
```suggestion
    pub value: Ether,
```
If we have this type then it makes sense to use here too.